Env(name="htseq", base="bioconda", packages="htseq>0.13")

with Stage("count_htseq"):
    rule htseq_count:
        message:
            "Counting per gene reads with htseq-count"
        input:
            bam = "{:prev:}/{:target:}.sorted.bam",
            gtf = "{:prev:}/{:target:}.gtf"
        output:
            counts = "{:this:}/{target}.htseq_counts",
        log:
            "{:this:}/{target}.log"
        params:
            max_reads_in_buffer = 30000000,  # 30m
            stranded = "reverse", # yes, no, reverse
            minaqual = 20,
            mode = "intersection-nonempty",
            nonunique = "none",
        threads:
            1  ## like fastqc, only 1 thread per file
        conda:
            "htseq"
        shell:
            "exec >/dev/null 2>&1;"
            "htseq-count"
            " --nprocesses={threads}"
            " --format=bam"
            " --order=pos"
            " --max-reads-in-buffer={params.max_reads_in_buffer}"
            " --stranded={params.stranded}"
            " -a={params.minaqual}"
            # --type=exon
            # --idattr=gene_id
            " --mode={params.mode}"
            " --nonunique={params.nonunique}"
            " {input.bam}"
            " {input.gtf}"
            " >{output.counts}"
